### PR TITLE
Reapply `mixed_semi_join` refactoring and bug fixes

### PR DESCRIPTION
**Pull request by mhaseeb123**
## Description
This PR reapplies changes from #16230 and adds bug fixes and performance improvements for mixed_semi_join.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/tests/join/mixed_join_tests.cu [hunk 1]
```diff
@@ -778,6 +778,131 @@ TYPED_TEST(MixedLeftSemiJoinTest, BasicEquality)
...
+
+  this->test({{2, 3, 9, 0, 1, 7, 4, 6, 5, 8}, {1, 2, 3, 4, 5, 6, 7, 8, 9, 0}},
+             {{6, 5, 9, 8, 10, 32}, {0, 1, 2, 3, 4, 5}, {7, 8, 9, 0, 1, 2}},
+             {0},
+             {1},
+             left_one_greater_right_one,
+             {2, 7, 8});
+}
+
+TYPED_TEST(MixedLeftSemiJoinTest, MixedLeftSemiJoinGatherMapLarge)
```
Comment (mhaseeb123) [inline cpp/tests/join/mixed_join_tests.cu:796]: New gtest that reproduces the bug that Spark reported.

File: cpp/src/join/mixed_join_kernels_semi.cu [hunk 2]
```diff
@@ -52,24 +56,27 @@ CUDF_KERNEL void __launch_bounds__(block_size)
...
+  auto equality              = single_expression_equality<has_nulls>{
+    evaluator, thread_intermediate_storage, swap_tables, equality_probe};
+
+  // Create set ref with the new equality comparator.
+  auto const set_ref_equality = set_ref.with_key_eq(equality);
 
-    left_table_keep_mask[outer_row_index] =
-      hash_table_view.contains(outer_row_index, hash_probe, equality);
+  // Find all the rows in the left table that are in the hash table.
+  for (auto outer_row_index = cudf::detail::grid_1d::global_thread_id<block_size>() / cg_size;
```
Comment (mhaseeb123) [inline cpp/src/join/mixed_join_kernels_semi.cu:75]: The bug: This was a simple `if` before so only querying `nelems/cg_size` items instead of all `nelems`

**Comment by mhaseeb123** [comment]
# Performance

## Benchmark
```bash
JOIN_NVBENCH -d 0 --benchmark mixed_left_semi_join --timeout 100
```

## Hardware
```
GPU: NVIDIA RTX 5880 Ada Generation
SM Version: 890 (PTX Version: 860)
Number of SMs: 110
SM Default Clock Rate: 18446744071874 MHz
Global Memory: 23879 MiB Free / 48632 MiB Total
Global Memory Bus Peak: 960 GB/sec (384-bit DDR @10001MHz)
Max Shared Memory: 100 KiB/SM, 48 KiB/Block
L2 Cache Size: 98304 KiB
Maximum Active Blocks: 24/SM
Maximum Active Threads: 1536/SM, 1024/Block
Available Registers: 65536/SM, 65536/Block
ECC Enabled: No
```

## Comparison Table
```
| Key | Nullable | left_size | right_size | GPU_Time_old | GPU_Time_new |
|-----|----------|-----------|------------|--------------|--------------|
| I32 |        0 |      1000 |       1000 |    94.260 us |    96.297 us |
| I32 |        0 |    100000 |       1000 |   132.938 us |   147.642 us |
| I32 |        0 |  10000000 |       1000 |     1.502 ms |     1.813 ms |
| I32 |        0 |    100000 |     100000 |   157.627 us |   141.354 us |
| I32 |        0 |  10000000 |     100000 |     1.479 ms |     1.685 ms |
| I32 |        0 |  10000000 |   10000000 |     5.867 ms |     4.158 ms |
| I32 |        1 |      1000 |       1000 |   108.833 us |    99.410 us |
| I32 |        1 |    100000 |       1000 |   143.874 us |   145.811 us |
| I32 |        1 |  10000000 |       1000 |     1.283 ms |     1.421 ms |
| I32 |        1 |    100000 |     100000 |   159.061 us |   145.652 us |
| I32 |        1 |  10000000 |     100000 |     1.156 ms |     1.426 ms |
| I32 |        1 |  10000000 |   10000000 |     2.951 ms |     1.899 ms |
| I64 |        0 |      1000 |       1000 |    94.945 us |    85.983 us |
| I64 |        0 |    100000 |       1000 |   134.796 us |   137.989 us |
| I64 |        0 |  10000000 |       1000 |     1.466 ms |     1.681 ms |
| I64 |        0 |    100000 |     100000 |   159.470 us |   142.163 us |
| I64 |        0 |  10000000 |     100000 |     1.510 ms |     1.720 ms |
| I64 |        0 |  10000000 |   10000000 |     6.862 ms |     5.228 ms |
| I64 |        1 |      1000 |       1000 |   106.791 us |    95.521 us |
| I64 |        1 |    100000 |       1000 |   143.220 us |   143.245 us |
| I64 |        1 |  10000000 |       1000 |     1.118 ms |     1.494 ms |
| I64 |        1 |    100000 |     100000 |   159.526 us |   146.640 us |
| I64 |        1 |  10000000 |     100000 |     1.178 ms |     1.428 ms |
| I64 |        1 |  10000000 |   10000000 |     2.980 ms |     1.988 ms |
```

File: cpp/src/join/mixed_join_semi.cu [hunk 3]
```diff
@@ -191,39 +145,48 @@ std::unique_ptr<rmm::device_uvector<size_type>> mixed_join_semi(
...
+    {},
+    cudf::detail::cuco_allocator<char>{rmm::mr::polymorphic_allocator<char>{}, stream},
+    {stream.value()}};
+
+  auto iter = thrust::make_counting_iterator(0);
 
   // skip rows that are null here.
   if ((compare_nulls == null_equality::EQUAL) or (not nullable(build))) {
-    hash_table.insert(iter, iter + right_num_rows, hash_build, equality_build, stream.value());
+    row_set.insert_async(iter, iter + right_num_rows, stream.value());
```
Comment (mhaseeb123) [inline cpp/src/join/mixed_join_semi.cu:169]: Some perf improvement through use of `_async` APIs.

File: cpp/src/join/mixed_join_common_utils.cuh [hunk 4]
```diff
@@ -160,6 +161,39 @@ struct pair_expression_equality : public expression_equality<has_nulls> {
...
+  {
+    using experimental::row::lhs_index_type;
+    using experimental::row::rhs_index_type;
+
+    return equality_comparator(lhs_index_type{lhs_row_index}, rhs_index_type{rhs_row_index}) &&
+           conditional_comparator(lhs_index_type{lhs_row_index}, rhs_index_type{rhs_row_index});
+  }
+};
+
+// A CUDA Cooperative Group of 4 threads for the hash set.
```
Comment (mhaseeb123) [inline cpp/src/join/mixed_join_common_utils.cuh:181]:
```suggestion
// A CUDA Cooperative Group of 1 thread for the hash set.
```

File: cpp/src/join/mixed_join_semi.cu [hunk 5]
```diff
@@ -191,39 +145,48 @@ std::unique_ptr<rmm::device_uvector<size_type>> mixed_join_semi(
...
-    hash_table.insert_if(
-      iter, iter + right_num_rows, stencil, pred, hash_build, equality_build, stream.value());
+    row_set.insert_if_async(iter, iter + right_num_rows, stencil, pred, stream.value());
   }
 
-  auto hash_table_view = hash_table.get_device_view();
-
-  detail::grid_1d const config(outer_num_rows, DEFAULT_JOIN_BLOCK_SIZE);
-  auto const shmem_size_per_block = parser.shmem_per_thread * config.num_threads_per_block;
+  detail::grid_1d const config(outer_num_rows * hash_set_type::cg_size, DEFAULT_JOIN_BLOCK_SIZE);
```
Comment (mhaseeb123) [inline cpp/src/join/mixed_join_semi.cu:180]: Bug fix (part 2): We need to launch `outer_num_rows * hash_set_type::cg_size` threads now

**Comment by mhaseeb123** [comment]
Performance results from Spark (credits: @zpuller).

## spark2a
```
Regression alerts
-----------------
--------------------------------------------------------------------
Name = query85
Means = 2845.4, 2253.0
Time diff = 592.4000000000001
Speedup = 1.2629383044829117
T-Test (test statistic, p value, df) = 2.43033039217987, 0.041181012901009915, 8.0
T-Test Confidence Interval = 30.304881537668848, 1154.4951184623314
ALERT: significant change has been detected (p-value < 0.05)
ALERT: improvement in performance has been observed
--------------------------------------------------------------------
Name = query90
Means = 1421.4, 863.4
Time diff = 558.0000000000001
Speedup = 1.6462821403752608
T-Test (test statistic, p value, df) = 2.6788276864789964, 0.02797678796082813, 8.0
T-Test Confidence Interval = 77.6591725781339, 1038.3408274218664
ALERT: significant change has been detected (p-value < 0.05)
ALERT: improvement in performance has been observed
--------------------------------------------------------------------
Name = query93
Means = 10904.4, 11764.0
Time diff = -859.6000000000004
Speedup = 0.9269296157769465
T-Test (test statistic, p value, df) = -2.9660190047403745, 0.017980126109303492, 8.0
T-Test Confidence Interval = -1527.9170780272916, -191.282921972709
ALERT: significant change has been detected (p-value < 0.05)
ALERT: regression in performance has been observed
```
## spark-h
```
====================
Results for query16
====================
Regression alerts
-----------------

Speedup results
-----------------
query16: Previous (6218.6 ms) vs Current (6250.6 ms) Diff -32 E2E 0.99x
====================
Results for query94
====================
Regression alerts
-----------------

Speedup results
-----------------
query94: Previous (5255.2 ms) vs Current (5044.4 ms) Diff 210 E2E 1.04x
```

File: cpp/src/join/mixed_join_kernels_semi.cu [hunk 6]
```diff
@@ -38,38 +38,48 @@ CUDF_KERNEL void __launch_bounds__(block_size)
...
+  auto const cg_grid_stride = cudf::detail::grid_1d::grid_stride<block_size>() / cg_size;
 
-    left_table_keep_mask[outer_row_index] =
-      hash_table_view.contains(outer_row_index, hash_probe, equality);
+  // Find all the rows in the left table that are in the hash table
+  for (auto outer_row_index = cudf::detail::grid_1d::global_thread_id<block_size>() / cg_size;
+       outer_row_index < outer_num_rows;
+       outer_row_index += cg_grid_stride) {
+    auto const result = set_ref_equality.contains(tile, outer_row_index);
+    if (tile.thread_rank() == 0) left_table_keep_mask[outer_row_index] = result;
```
Comment (PointKernel) [inline cpp/src/join/mixed_join_kernels_semi.cu:82]:
```suggestion
    if (tile.thread_rank() == 0) { left_table_keep_mask[outer_row_index] = result; }
```
nit: _always braces_, even for a single line if/while/for

File: cpp/tests/join/mixed_join_tests.cu (hunk 1 above)
Comment (PointKernel) [reply cpp/tests/join/mixed_join_tests.cu:796]: To test the scenario not covered by the current implementation, we need an input where the number of rows exceeds the block size. Given this condition, the current test seems to be overly complex. Is there a way we can simplify the test a bit?

Comment (mhaseeb123) [reply cpp/tests/join/mixed_join_tests.cu:796]: Separating the boilerplate of random data and nullmask generation, the test simply compares the number of rows in the output from `mixed_semi_join` and `conditional_left_semi_join` with an identical filter. I was actually thinking of comparing the returned columns themselves as well 😅

Comment (mhaseeb123) [reply cpp/tests/join/mixed_join_tests.cu:796]: Or I can modify the test at L781 to use larger input tables and use a ground truth expected vector.

Comment (PointKernel) [reply cpp/tests/join/mixed_join_tests.cu:796]: https://github.com/rapidsai/cudf/blob/dc57c1b1284816d0e5ed7493e6b661590c305511/cpp/tests/join/mixed_join_tests.cu#L64-L125

could we reuse those random data generators for this test case?

Comment (mhaseeb123) [reply cpp/tests/join/mixed_join_tests.cu:796]: Oh yes, I didn't even see this hence all this effort. Let me use these instead

File: cpp/src/join/mixed_join_common_utils.cuh [hunk 7]
```diff
@@ -160,6 +161,39 @@ struct pair_expression_equality : public expression_equality<has_nulls> {
...
+auto constexpr DEFAULT_MIXED_SEMI_JOIN_CG_SIZE = 1;
```
Comment (mythrocks) [inline cpp/src/join/mixed_join_common_utils.cuh:182]: I might not be reading this correctly.  Does the size in the comment (4) contradict that specified in the constant definition (1)?

File: cpp/src/join/mixed_join_kernels_semi.cu [hunk 8]
```diff
@@ -38,38 +38,48 @@ CUDF_KERNEL void __launch_bounds__(block_size)
...
+
+  auto const tile = cg::tiled_partition<cg_size>(cg::this_thread_block());
+
   // Normally the casting of a shared memory array is used to create multiple
   // arrays of different types from the shared memory buffer, but here it is
   // used to circumvent conflicts between arrays of different types between
   // different template instantiations due to the extern specifier.
   extern __shared__ char raw_intermediate_storage[];
-  cudf::ast::detail::IntermediateDataType<has_nulls>* intermediate_storage =
+  auto intermediate_storage =
```
Comment (mythrocks) [inline cpp/src/join/mixed_join_kernels_semi.cu:55]: 👍

File: cpp/src/join/mixed_join_common_utils.cuh (hunk 7 above)
Comment (mythrocks) [reply cpp/src/join/mixed_join_common_utils.cuh:182]: Actually, should this not be set to `4` instead of `1`?

Comment (mhaseeb123) [reply cpp/src/join/mixed_join_common_utils.cuh:182]: Recently updated `cg_size` from 4 to 1. Thanks for pointing out. Will update the comment as well.

File: cpp/src/join/mixed_join_kernels_semi.cu [hunk 9]
```diff
@@ -38,38 +38,48 @@ CUDF_KERNEL void __launch_bounds__(block_size)
...
+  // Grid stride for the tile
+  auto const cg_grid_stride = cudf::detail::grid_1d::grid_stride<block_size>() / cg_size;
 
-    left_table_keep_mask[outer_row_index] =
-      hash_table_view.contains(outer_row_index, hash_probe, equality);
+  // Find all the rows in the left table that are in the hash table
+  for (auto outer_row_index = cudf::detail::grid_1d::global_thread_id<block_size>() / cg_size;
+       outer_row_index < outer_num_rows;
+       outer_row_index += cg_grid_stride) {
+    auto const result = set_ref_equality.contains(tile, outer_row_index);
```
Comment (mythrocks) [inline cpp/src/join/mixed_join_kernels_semi.cu:81]: I'm fairly certain I'm not understanding this correctly.  (Or maybe I'm not understanding how CGs work.) Please ELI5 the following:

Consider `cg_size == 4`.  The `outer_row_index` would be the same for all the threads in the cooperative group.  Wouldn't the same `result` for the same `cg` be calculated 4 times?  

I haven't grasped how `result` changes here, for threads 0-3, for the same CG.  It looks to me like the calculation is done 4 times, and then the first one (`thread_rank() == 0) writes out the result.

Comment (mythrocks) [reply cpp/src/join/mixed_join_kernels_semi.cu:81]: P.S. Apologies if this should be obvious.  I'm not seeing it yet.

Comment (mhaseeb123) [reply cpp/src/join/mixed_join_kernels_semi.cu:81]: In your above example, all 4 threads in a tile of `cg_size=4` would collaborate (in probing the static_set) to find the result. Once all four threads in the `tile` have the same result, we only want one of them (here the one with rank = 0 within the tile) to actually write it out.

File: cpp/tests/join/mixed_join_tests.cu (hunk 1 above)
Comment (mhaseeb123) [reply cpp/tests/join/mixed_join_tests.cu:796]: All done!

File: cpp/tests/join/mixed_join_tests.cu [hunk 10]
```diff
@@ -778,6 +778,138 @@ TYPED_TEST(MixedLeftSemiJoinTest, BasicEquality)
...
+{
+  using T1 = double;
+
+  // Column size
```
Comment (mhaseeb123) [inline cpp/tests/join/mixed_join_tests.cu:800]:
```suggestion
  // Number of rows in each column
```

File: cpp/tests/join/mixed_join_tests.cu (hunk 1 above)
Comment (PointKernel) [reply cpp/tests/join/mixed_join_tests.cu:796]: @mhaseeb123 Thanks for your quick turnaround on the comments!

**Review by PointKernel — APPROVED** [review]
Looks great :rocket:

File: cpp/src/join/mixed_join_kernels_semi.cu (hunk 9 above)
Comment (mythrocks) [reply cpp/src/join/mixed_join_kernels_semi.cu:81]: Ah, sorry.  Ok, I think I understand.  I should have paid attention to the fact that the `tile` is being sent to `set.contains()`.  I was focused on the `outer_row_index`.

**Review by mythrocks — APPROVED** [review]
This was an enlightening read.  Very clean code, too.  Thanks, @mhaseeb123.

LGTM.

**Review by ttnghia — APPROVED** [review]
